src/compiler/transformers/es2017.ts

```diff
@@ -83,6 +85,103 @@ namespace ts {
             }
         }
 
+        function asyncBodyVisitor(node: Node): VisitResult<Node> {
+            switch (node.kind) {
+                case SyntaxKind.VariableStatement:
+                    return visitVariableStatementInAsyncBody(<VariableStatement>node);
+                case SyntaxKind.ForStatement:
+                    return visitForStatementInAsyncBody(<ForStatement>node);
+                case SyntaxKind.ForInStatement:
+                    return visitForInStatementInAsyncBody(<ForInStatement>node);
+                case SyntaxKind.ForOfStatement:
+                    return visitForOfStatementInAsyncBody(<ForOfStatement>node);
+                case SyntaxKind.Block:
+                case SyntaxKind.SwitchStatement:
+                case SyntaxKind.CaseBlock:
+                case SyntaxKind.CaseClause:
+                case SyntaxKind.DefaultClause:
+                case SyntaxKind.TryStatement:
+                case SyntaxKind.DoStatement:
+                case SyntaxKind.WhileStatement:
+                case SyntaxKind.IfStatement:
+                case SyntaxKind.WithStatement:
+                    return visitEachChild(node, asyncBodyVisitor, context);
+                case SyntaxKind.CatchClause:
+                    return visitCatchClauseInAsyncBody(<CatchClause>node);
+            }
+            return visitor(node);
+        }
+
+        function visitCatchClauseInAsyncBody(node: CatchClause) {
+            const catchClauseNames = createUnderscoreEscapedMap<true>();
+            recordDeclarationName(node.variableDeclaration, catchClauseNames);

```

Comment:
variableDeclaration may be undefined

Reply:
Actually, by the time we get to the `es2017` transform a `CatchClause` **must** have a catch variable, otherwise it's not valid ES2017. Assigning the catch variable is handled in the `esnext` transform before the `es2017` transform is run.

Reply:
I've added a test case in e655446 to verify this case.